posthog/models/cohort.py

```diff
@@ -102,29 +102,28 @@ def get_analytics_metadata(self):
             "deleted": self.deleted,
         }
 
-    def calculate_people(self):
+    def calculate_people(self, batch_size=50000, pg_batch_size=5000):
         if self.is_static:
             return
         try:
-            persons_query = self._clickhouse_persons_query()
+            cursor = 0
+            # increase by 1 to test pagination
+            persons_query = self._clickhouse_persons_query(batch_size=batch_size, offset=0)
 
             try:
-                sql, params = persons_query.distinct("pk").only("pk").query.sql_with_params()
+                persons = persons_query.distinct("pk")
             except EmptyResultSet:
-                query = DELETE_QUERY.format(cohort_id=self.pk)
-                params = {}
+                CohortPeople.objects.filter(cohort_id=self.pk).delete()
             else:
-                query = f"""
-                    {DELETE_QUERY};
-                    {UPDATE_QUERY};
-                """.format(
-                    cohort_id=self.pk,
-                    values_query=sql.replace('FROM "posthog_person"', f', {self.pk} FROM "posthog_person"', 1,),
-                )
+                with transaction.atomic():
+                    CohortPeople.objects.filter(cohort_id=self.pk).delete()
+                    while persons:
+                        to_insert = [CohortPeople(person_id=p.pk, cohort_id=self.pk) for p in persons]
+                        CohortPeople.objects.bulk_create(to_insert, batch_size=pg_batch_size)
+
+                        cursor += batch_size
+                        persons = self._clickhouse_persons_query(batch_size=batch_size, offset=cursor)
```

Comment:
Is using an offset here going to cause consistency issues? i.e. with the results set changing under our feet as we paginate. Could or do we freeze the set by filtering less than a specific date on the events table? And likewise on the persons table.

It's probably not the most important consistency issue in the grand scheme but thought worth mentioning.

Reply:
(anyway, if it's possible, we should add some sorting on something strictly increasing and use that as the cursor.)